evmcore/tx_validation.go

```diff
@@ -40,78 +50,61 @@ type validationOptions struct {
 	eip7623 bool // Fork indicator whether we are using EIP-7623 floor gas validation.
 	eip7702 bool // Fork indicator whether we are using EIP-7702 set code transactions.
 
-	currentState   TxPoolStateDB // Current state in the blockchain head
-	currentMaxGas  uint64        // Current gas limit for transaction caps
-	currentBaseFee *big.Int      // Current base fee for transaction caps
-	minTip         *big.Int      // Minimum gas tip to enforce for acceptance into the pool
+	currentMaxGas  uint64   // Current gas limit for transaction caps
+	currentBaseFee *big.Int // Current base fee for transaction caps
 
-	locals  *accountSet // Set of local transaction to exempt from eviction rules
-	isLocal bool        // Whether the transaction came from a local source
-
-	signer types.Signer
+	signer types.Signer // sender is included as part of network rules because it depends on the chainId.
 }
 
 // validateTx checks whether a transaction is valid according to the current
 // options and adheres to some heuristic limits of the local node (price and size).
-func validateTx(tx *types.Transaction, opt validationOptions) error {
+func validateTx(
+	tx *types.Transaction,
+	opt validationOptions,
+	netRules NetworkRulesForValidateTx) error {
 
-	// Accept only legacy transactions until EIP-2718/2930 activates.
-	// Since both eip-2718 and eip-2930 are activated in the berlin fork
-	// (https://blog.ethereum.org/2021/03/08/ethereum-berlin-upgrade-announcement),
-	// they can be grouped in a single flag.
-	if !opt.eip2718 && tx.Type() != types.LegacyTxType {
-		return ErrTxTypeNotSupported
+	if err := ValidateTxStatic(tx); err != nil {
+		return err
 	}
-	// Reject dynamic fee transactions until EIP-1559 activates.
-	if !opt.eip1559 && tx.Type() == types.DynamicFeeTxType {
-		return ErrTxTypeNotSupported
+
+	from, err := ValidateTxForNetworkRules(tx, netRules)
+	if err != nil {
+		return err
 	}
-	// Reject blob transactions until EIP-4844 activates or if is already EIP-4844 and they are not empty
-	if tx.Type() == types.BlobTxType {
-		if !opt.eip4844 {
-			return ErrTxTypeNotSupported
-		}
-		// For now, Sonic only supports Blob transactions without blob data.
-		if len(tx.BlobHashes()) > 0 ||
-			(tx.BlobTxSidecar() != nil && len(tx.BlobTxSidecar().BlobHashes()) > 0) {
-			return ErrTxTypeNotSupported
-		}
+
+	if err := validateTxForPool(tx, opt, from); err != nil {
+		return err
 	}
-	// validate EIP-7702 transactions, part of prague revision
-	if tx.Type() == types.SetCodeTxType {
-		// Check minimum revision
-		if !opt.eip7702 {
-			return ErrTxTypeNotSupported
-		}
 
-		// Check non-empty authorization list
-		if len(tx.SetCodeAuthorizations()) == 0 {
-			return ErrEmptyAuthorizations
-		}
+	// leave state accesses for last check
```

Comment:
nit: not a very useful comment ..

Reply:
https://github.com/0xsoniclabs/sonic/pull/200/commits/32055b804e8a5906eb9786572ca27dc9a38ca183